File: app/src/main/java/com/khaled/move/navigation/foot/eta/EtaCalculator.kt
package com.khaled.move.navigation.foot.eta

import java.time.Clock
import java.time.Instant

class EtaCalculator(
    private val clock: Clock = Clock.systemDefaultZone(),
) {
    private var previousRemainingSeconds: Double? = null

    fun reset() {
        previousRemainingSeconds = null
    }

    fun calculate(remainingDistanceMeters: Double, speedMetersPerSecond: Double): EtaEstimate {
        val effectiveSpeed = when {
            remainingDistanceMeters <= 0.0 -> 0.0
            speedMetersPerSecond < 0.2 -> STATIONARY_EFFECTIVE_SPEED_METERS_PER_SECOND
            else -> speedMetersPerSecond.coerceIn(
                MIN_WALKING_SPEED_METERS_PER_SECOND,
                MAX_WALKING_SPEED_METERS_PER_SECOND
            )
        }
        val rawRemainingSeconds = if (remainingDistanceMeters <= 0.0) 0.0 else remainingDistanceMeters / effectiveSpeed
        val previous = previousRemainingSeconds
        val remainingSeconds = if (previous == null || remainingDistanceMeters <= 0.0) {
            rawRemainingSeconds
        } else {
            val delta = rawRemainingSeconds - previous
            if (kotlin.math.abs(delta) < ETA_DEADBAND_SECONDS) {
                previous
            } else {
                previous + delta * ETA_SMOOTHING_ALPHA
            }
        }
        previousRemainingSeconds = remainingSeconds
        return EtaEstimate(
            remainingDurationSeconds = remainingSeconds,
            arrivalTime = Instant.ofEpochMilli(clock.millis() + (remainingSeconds * 1000).toLong()),
        )
    }

    private companion object {
        const val MIN_WALKING_SPEED_METERS_PER_SECOND = 0.35
        const val MAX_WALKING_SPEED_METERS_PER_SECOND = 2.35
        const val STATIONARY_EFFECTIVE_SPEED_METERS_PER_SECOND = 0.65
        const val ETA_DEADBAND_SECONDS = 18.0
        const val ETA_SMOOTHING_ALPHA = 0.35
    }
}

data class EtaEstimate(
    val remainingDurationSeconds: Double,
    val arrivalTime: Instant,
)

File: app/src/main/cpp/engine/EtaCalculator.cpp
#include "EtaCalculator.h"
#include <cmath>
#include <algorithm>

namespace move {

EtaCalculator::EtaCalculator() : previousRemainingSeconds(0.0), hasPrevious(false) {}

void EtaCalculator::reset() {
    hasPrevious = false;
}

EtaEstimate EtaCalculator::calculate(double remainingDistanceMeters, double speedMetersPerSecond, long long currentTimestampMillis) {
    double effectiveSpeed;
    if (remainingDistanceMeters <= 0.0) {
        effectiveSpeed = 0.0;
    } else if (speedMetersPerSecond < 0.2) {
        effectiveSpeed = STATIONARY_EFFECTIVE_SPEED;
    } else {
        effectiveSpeed = std::max(MIN_WALKING_SPEED, std::min(MAX_WALKING_SPEED, speedMetersPerSecond));
    }

    double rawRemainingSeconds = (remainingDistanceMeters <= 0.0) ? 0.0 : remainingDistanceMeters / effectiveSpeed;

    double remainingSeconds;
    if (!hasPrevious || remainingDistanceMeters <= 0.0) {
        remainingSeconds = rawRemainingSeconds;
    } else {
        double delta = rawRemainingSeconds - previousRemainingSeconds;
        if (std::abs(delta) < ETA_DEADBAND_SECONDS) {
            remainingSeconds = previousRemainingSeconds;
        } else {
            remainingSeconds = previousRemainingSeconds + delta * ETA_SMOOTHING_ALPHA;
        }
    }

    previousRemainingSeconds = remainingSeconds;
    hasPrevious = true;

    return {
        remainingSeconds,
        currentTimestampMillis + static_cast<long long>(remainingSeconds * 1000.0)
    };
}

} // namespace move
